### PR TITLE
Feature/model loader refactor

### PR DESCRIPTION
**Pull request by luuk234**
### Description

alright this is a big one but most of the changes are due to renaming a couple types.

there are also still some things to do but this is probably better for another pr since there are dependencies on this task

things todo:
- better organisation of cpu/gpu types, ie files.
- there are still some gpu dependencies in cpu structs but this shouldn't be a problem for now
- profile to see how much more expensive model loading is now, and perhapse add shared_pointers to some of the  cpu types.
- write unit tests 

things done:
- seperated cpu and gpu versions of models, materials, meshes and images.
- cleaned up model loader interface and removed graphics/ resource manager dependencies from model loader
- added model resource manager
- other small things

### Issues

- https://bubonic-brotherhood.codecks.io/decks/7-engine/card/1a7-model-loader-refactor)

### Test criteria

- test if visual output is see same

### CODE REVIEW

File: engine/source/scene_loader.cpp [hunk 1]
```diff
@@ -12,40 +12,42 @@
...
         RelationshipHelpers::AttachChild(ecs.registry, entity, childEntity);
     }
 
     return entity;
 }
+
+std::vector<entt::entity> SceneLoading::LoadModelIntoECSAsHierarchy(ECS& ecs, const GPUModel& model)
+{
+    std::vector<entt::entity> entities {};
+    std::cout << "  dfrster 8" << std::endl;
```
Comment (dumitrurares14) [inline engine/source/scene_loader.cpp:45]: Remove testing cout

File: engine/source/scene_loader.cpp [hunk 2]
```diff
@@ -12,40 +12,42 @@
...
-
-    return entities;
-}
-
-entt::entity SceneLoader::LoadNodeRecursive(const std::shared_ptr<GraphicsContext>& context, ECS& ecs, const Hierarchy::Node& currentNode)
+entt::entity LoadNodeRecursive(ECS& ecs, const Hierarchy::Node& currentNode, const GPUModel& model)
 {
     const entt::entity entity = ecs.registry.create();
 
+    std::cout << "sdfadw" << std::endl;
```
Comment (dumitrurares14) [inline engine/source/scene_loader.cpp]: remove

File: engine/renderer/private/renderer.cpp [hunk 3]
```diff
@@ -30,10 +30,13 @@
 #include "resource_management/buffer_resource_manager.hpp"
 #include "resource_management/image_resource_manager.hpp"
 #include "resource_management/material_resource_manager.hpp"
+#include "resource_management/model_resource_manager.hpp"
 #include "single_time_commands.hpp"
 #include "vulkan_context.hpp"
 #include "vulkan_helper.hpp"
 
+#include <resource_management/mesh_resource_manager.hpp>
```
Comment (MuskettaMan) [inline engine/renderer/private/renderer.cpp:38]: should be quotes

Comment (MuskettaMan) [reply engine/renderer/private/renderer.cpp:38]: and doesnt need a whiteline

File: engine/renderer/private/renderer.cpp [hunk 4]
```diff
@@ -49,12 +52,13 @@ Renderer::Renderer(ApplicationModule& application, const std::shared_ptr<Graphic
     InitializeBloomTargets();
     LoadEnvironmentMap();
 
-    _modelLoader = std::make_unique<ModelLoader>(_context, _ecs);
+    _modelLoader = std::make_unique<ModelLoader>();
 
     _batchBuffer = std::make_shared<BatchBuffer>(_context, 256 * 1024 * 1024, 256 * 1024 * 1024);
+    context->Resources()->MeshResourceManager().SetBatchBuffer(_batchBuffer);
```
Comment (MuskettaMan) [inline engine/renderer/private/renderer.cpp:58]: Hmm, I would think it makes more sense to pass the batch buffer in the create call to a mesh

File: engine/renderer/private/renderer_module.cpp [hunk 5]
```diff
@@ -19,7 +19,8 @@ RendererModule::RendererModule()
 ModuleTickOrder RendererModule::Init(Engine& engine)
 {
     auto ecs = engine.GetModule<OldEngine>().GetECS();
-    _context = std::make_shared<GraphicsContext>(engine.GetModule<ApplicationModule>().GetVulkanInfo());
+    auto s = engine.GetModule<ApplicationModule>().GetVulkanInfo();
```
Comment (MuskettaMan) [inline engine/renderer/private/renderer_module.cpp:22]: s isnt a great variable name

File: engine/renderer/private/resource_management/mesh_resource_manager.cpp [hunk 6]
```diff
@@ -1,8 +1,57 @@
...
 #include "gpu_resources.hpp"
+#include "single_time_commands.hpp"
+#include "vulkan_context.hpp"
 
 template <>
-std::weak_ptr<ResourceManager<Mesh>> ResourceHandle<Mesh>::manager = {};
+std::weak_ptr<ResourceManager<GPUMesh>> ResourceHandle<GPUMesh>::manager = {};
 
-MeshResourceManager::MeshResourceManager() = default;
+ResourceHandle<GPUMesh> MeshResourceManager::Create(const CPUMesh& cpuMesh, std::vector<ResourceHandle<GPUMaterial>> materials)
```
Comment (MuskettaMan) [inline engine/renderer/private/resource_management/mesh_resource_manager.cpp:11]: vector should be passed as const ref

File: engine/renderer/private/resource_management/model_resource_manager.cpp [hunk 7]
```diff
@@ -0,0 +1,65 @@
+#include "resource_management/model_resource_manager.hpp"
+
+#include "batch_buffer.hpp"
+#include "model.hpp"
+#include "resource_management/image_resource_manager.hpp"
+#include "resource_management/material_resource_manager.hpp"
+#include "resource_management/mesh_resource_manager.hpp"
+
+#include "fastgltf/base64.hpp"
```
Comment (MuskettaMan) [inline engine/renderer/private/resource_management/model_resource_manager.cpp:9]: required?

File: engine/renderer/private/resource_management/model_resource_manager.cpp [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+{
+    GPUModel model;
+
+    for (const auto& i : data.textures)
+    {
+
+        model.textures.emplace_back(_imageResourceManager->Create(i));
+    }
+
+    for (const auto& i : data.materials)
```
Comment (MuskettaMan) [inline engine/renderer/private/resource_management/model_resource_manager.cpp:32]: i is not a good variable name

File: engine/renderer/private/resource_management/model_resource_manager.cpp [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    , _meshResourceManager(meshResourceManager)
+{
+}
+
+ResourceHandle<GPUModel>
+ModelResourceManager::Create(const CPUModel& data)
+{
+    GPUModel model;
+
+    for (const auto& i : data.textures)
```
Comment (MuskettaMan) [inline engine/renderer/private/resource_management/model_resource_manager.cpp:26]: i is not a good variable name

File: engine/renderer/private/resource_management/model_resource_manager.cpp [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+
+        material.occlusionStrength = i.occlusionStrength;
+        material.occlusionUVChannel = i.occlusionUVChannel;
+        if (i.occlusionMap.has_value())
+            material.occlusionMap = model.textures[i.occlusionMap.value()];
+
+        model.materials.emplace_back(_materialResourceManager->Create(material));
+    }
+
+    for (const auto& i : data.meshes)
```
Comment (MuskettaMan) [inline engine/renderer/private/resource_management/model_resource_manager.cpp:58]: i is not a good variable name

File: engine/renderer/public/gpu_resources.hpp [hunk 11]
```diff
@@ -231,7 +231,8 @@ struct Mesh
         uint32_t indexOffset;
         float boundingRadius;
 
-        ResourceHandle<Material> material;
+        // Todo: change this into index into mesh material slots?
```
Comment (MuskettaMan) [inline engine/renderer/public/gpu_resources.hpp:234]: I think this should stay a resource handle, i think the indices are more error prone

**Review by MuskettaMan — CHANGES_REQUESTED** [review]
Some questions and things to changes, but overall this looks really good. Much clearer where what resources are. Especially using CPU variants into the GPU ones makes a lot of sense.

File: engine/renderer/private/mesh_primitives.cpp [hunk 12]
```diff
@@ -8,16 +8,16 @@ void AddTriangle(std::vector<uint32_t>& indices, std::array<uint32_t, 3> triangl
...
-    StagingMesh::Primitive primitive;
+    CPUMesh::Primitive primitive;
 
     uint32_t totalVertices = 2 + (stacks - 1) * slices;
 
     primitive.vertices.reserve(totalVertices);
 
     // TODO: Consider this based on the total amount of indices instead.
-    primitive.materialIndex = std::nullopt;
+    primitive.materialIndex = 0;
```
[the remark this answers is not in the document]
Comment (luuk234) [reply engine/renderer/private/mesh_primitives.cpp:20]: materialIndex is now just a uint32_t not optional anymore

**Review by MuskettaMan — APPROVED** [review]

**Pull request merged**